src/vs/workbench/contrib/chat/common/promptSyntax/service/promptsServiceImpl.ts

```diff
@@ -154,16 +154,111 @@ export class PromptsService extends Disposable implements IPromptsService {
 			this.fileLocator.listFiles(type, PromptsStorage.local, token).then(uris => uris.map(uri => ({ uri, storage: PromptsStorage.local, type } satisfies ILocalPromptPath))),
-			this.getExtensionContributions(type)
+			this.getExtensionPromptFiles(type, token),
 		]);
 
 		return [...prompts.flat()];
 	}
 
+	/**
+	 * Registry of CustomAgentsProvider instances. Extensions can register providers via the proposed API.
+	 */
+	private readonly customAgentsProviders: Array<{
+		extension: IExtensionDescription;
+		onDidChangeCustomAgents?: Event<void>;
+		provideCustomAgents: (options: ICustomAgentQueryOptions, token: CancellationToken) => Promise<IExternalCustomAgent[] | undefined>;
+	}> = [];
+
+	/**
+	 * Registers a CustomAgentsProvider. This will be called by the extension host bridge when
+	 * an extension registers a provider via vscode.chat.registerCustomAgentsProvider().
+	 */
+	public registerCustomAgentsProvider(extension: IExtensionDescription, provider: {
+		onDidChangeCustomAgents?: Event<void>;
+		provideCustomAgents: (options: ICustomAgentQueryOptions, token: CancellationToken) => Promise<IExternalCustomAgent[] | undefined>;
+	}): IDisposable {
+		if (!isProposedApiEnabled(extension, 'chatParticipantPrivate')) {
+			throw new Error(`Extension '${extension.identifier.value}' CANNOT register CustomAgentsProvider. The 'chatParticipantPrivate' proposal must be enabled.`);
+		}
+
+		const providerEntry = { extension, ...provider };
+		this.customAgentsProviders.push(providerEntry);
+
+		const disposables = new DisposableStore();
+
+		// Listen to provider change events to rerun computeListPromptFiles
+		if (provider.onDidChangeCustomAgents) {
+			disposables.add(provider.onDidChangeCustomAgents(() => {
+				this.cachedFileLocations[PromptsType.agent] = undefined;
+				this.cachedCustomAgents.refresh();
+			}));
+		}
+
+		// Invalidate agent cache when providers change
+		this.cachedFileLocations[PromptsType.agent] = undefined;
+		this.cachedCustomAgents.refresh();
+
+		disposables.add({
+			dispose: () => {
+				const index = this.customAgentsProviders.findIndex((p) => p === providerEntry);
+				if (index >= 0) {
+					this.customAgentsProviders.splice(index, 1);
+					this.cachedFileLocations[PromptsType.agent] = undefined;
+					this.cachedCustomAgents.refresh();
+				}
+			}
+		});
+
+		return disposables;
+	}
+
+	private async listCustomAgentsFromProvider(token: CancellationToken): Promise<IPromptPath[]> {
+		const result: IPromptPath[] = [];
+
+		if (this.customAgentsProviders.length === 0) {
+			return result;
+		}
+
+		// Collect agents from all providers
+		for (const providerEntry of this.customAgentsProviders) {
+			try {
+				const agents = await providerEntry.provideCustomAgents({}, token);
+				if (!agents || token.isCancellationRequested) {
+					continue;
+				}
+
+				for (const agent of agents) {
+					try {
+						await this.filesConfigService.updateReadonly(agent.uri, true);
```

Comment:
Maybe in the future the provider can provide the information if its agents are modifiable